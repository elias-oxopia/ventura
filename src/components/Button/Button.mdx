---
name: Button
menu: Components
route: /button
---

import { Playground, Props } from 'docz'
import { Activity, Trash2 } from '../../index.ts'
import Button from './Button.tsx'

# Button

Buttons make common actions more obvious and help users more easily perform them. Buttons use labels and sometimes icons to communicate the action that will occur when the user touches them.

### Best practices

- Group buttons logically into sets based on usage and importance.
- Ensure that button actions are clear and consistent.
- The main action of a group set can be a primary button.
- Select a single button variation and do not mix them.

## Basic usage

<Playground>
  <Button>Click me</Button>
</Playground>

## Button types

<Playground>
  <Button type="primary">Click me</Button>
  <br/>
  <Button type="secondary">Click me</Button>
</Playground>

## Button disabled

<Playground>
  <Button isDisabled>Click me</Button>
</Playground>

## Button loading

<Playground>
  <Button isLoading>Click me</Button>
</Playground>

## Button with icon

<Playground>
  <Button suffixIcon={<Activity />}>Click me</Button>
  <br/>
  <Button prefixIcon={<Trash2 />}>Click me</Button>
</Playground>

## API

<Props of={Button} />
