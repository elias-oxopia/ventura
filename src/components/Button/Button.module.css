.button {
  border-radius: var(--border-radius-small);
  line-height: 14px;
  font-size: 14px;
  padding: 10px 14px;
  border: none;
  cursor: pointer;
  display: flex;
}

.smallButton {
  padding: 3.5px 10px;
}

.typePrimary:hover {
  background-color: var(--color-primary-hover);
}

.typeSecondary:hover {
  background-color: var(--color-secondary-hover);
}

.typePrimary:focus {
  outline: none;
  box-shadow: 0 0 0 2px var(--color-focus);
}

.typeSecondary:focus {
  outline: none;
  box-shadow: 0 0 0 2px var(--color-focus);
}

.typePrimary,
.typePrimary:active {
  background-color: var(--color-primary);
}

.typeSecondary,
.typeSecondary:active {
  background-color: var(--color-secondary);
}

.button:disabled {
  color: var(--color-disabled-font);
  background-color: var(--color-disabled);
  cursor: default;
}

.label {
  line-height: inherit;
  max-height: 14px;
}

.smallLabel {
  font-size: calc(var(--font-size) - 2px);
}

.labelWithPrefixIcon {
  margin-left: 8px;
}

.labelWithSuffixIcon {
  margin-right: 8px;
}

.spinner {
  width: 14px;
  height: 14px;
  vertical-align: -2px;
}
