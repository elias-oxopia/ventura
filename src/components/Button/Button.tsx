import React, { MouseEventHandler } from 'react';
import classNames from 'classnames';

import Spinner from '../utils/Spinner/Spinner';

import cssReset from '../../css-reset.module.css';
import styles from './Button.module.css';

interface Props {
  children: React.ReactNode;
  type?: 'primary' | 'secondary';
  onClick?: MouseEventHandler<HTMLButtonElement>;
  className?: string;
  isDisabled?: boolean;
  isLoading?: boolean;
  isSubmit?: boolean;
  prefixIcon?: React.ReactElement;
  suffixIcon?: React.ReactElement;
  name?: string;
}

const Button: React.FC<Props> = ({
  children,
  type = 'primary',
  onClick,
  className,
  isDisabled,
  isLoading,
  isSubmit,
  prefixIcon,
  suffixIcon,
  name,
}: Props) => {
  const buttonClassNames = classNames(
    cssReset.ventura,
    styles.button,
    {
      [styles.typePrimary]: type === 'primary',
      [styles.typeSecondary]: type === 'secondary',
    },
    className,
  );

  const labelClassNames = classNames({
    [styles.labelWithPrefixIcon]: Boolean(prefixIcon) || isLoading,
    [styles.labelWithSuffixIcon]: Boolean(suffixIcon),
  });

  return (
    /* eslint-disable-next-line react/button-has-type */
    <button
      disabled={isDisabled || isLoading}
      className={buttonClassNames}
      type={isSubmit ? 'submit' : 'button'}
      onClick={isLoading || suffixIcon ? undefined : onClick}
      data-testid={`button-${name}`}
    >
      {isLoading ? <Spinner className={styles.spinner} /> : prefixIcon}
      <span className={labelClassNames}>{children}</span>
      {isLoading || suffixIcon}
    </button>
  );
};

export default Button;
