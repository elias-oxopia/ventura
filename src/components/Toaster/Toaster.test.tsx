import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';

import Button from '../Button/Button';
import toaster from './Toaster';

describe('Toaster', () => {
  test('snapshot of error toast', () => {
    const { baseElement, getByTestId } = render(
      <Button
        onClick={() => {
          toaster.error('Message');
        }}
        name="toaster"
      >
        Trigger toaster
      </Button>,
    );

    const button = getByTestId('button-toaster');
    fireEvent.click(button);

    expect(baseElement).toMatchSnapshot();
  });

  test('snapshot of success toast', () => {
    const { baseElement, getByTestId } = render(
      <Button
        onClick={() => {
          toaster.success('Message');
        }}
        name="toaster"
      >
        Trigger toaster
      </Button>,
    );

    const button = getByTestId('button-toaster');
    fireEvent.click(button);

    expect(baseElement).toMatchSnapshot();
  });

  test('toast has correct message', () => {
    const { getByTestId, getByText } = render(
      <Button
        onClick={() => {
          toaster.error('Something went wrong.');
        }}
        name="toaster"
      >
        Trigger toaster
      </Button>,
    );

    const button = getByTestId('button-toaster');
    fireEvent.click(button);

    const toasterElement = getByText(/Something went wrong./);

    expect(toasterElement).toBeInTheDocument();
  });

  test('toast onClose', async () => {
    const { getByTestId, getByText } = render(
      <Button
        onClick={() => {
          toaster.success('I hope this closes.');
        }}
        name="toaster"
      >
        Trigger toaster
      </Button>,
    );

    const button = getByTestId('button-toaster');
    fireEvent.click(button);

    const toasterElement = getByText(/I hope this closes./);

    const closeIcon = toasterElement.parentElement?.parentElement?.querySelector('.closeAlert');
    fireEvent.click(closeIcon!);

    await waitFor(() => expect(toasterElement).not.toBeInTheDocument());

    expect(toasterElement).not.toBeInTheDocument();
  });
});
