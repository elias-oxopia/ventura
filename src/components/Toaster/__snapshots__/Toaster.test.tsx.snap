// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toaster snapshot of error toast 1`] = `
<body>
  <div
    id="ventura-toasters"
  >
    <span
      class="toastManagerContainer"
    >
      <div
        class="toastContainer"
        data-state="entering"
        style="height: 0px; margin-bottom: 0px;"
      >
        <div
          class="toastPadding"
        >
          <div
            class="ventura alert alertError"
          >
            <div
              class="contentContainer closable"
            >
              <svg
                class="icon iconError"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="10"
                />
                <line
                  x1="12"
                  x2="12"
                  y1="8"
                  y2="12"
                />
                <line
                  x1="12"
                  x2="12.01"
                  y1="16"
                  y2="16"
                />
              </svg>
              <span
                class="message"
              >
                Message
              </span>
            </div>
            <div
              class="closeContainer"
            >
              <svg
                class="closeAlert"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="18"
                  x2="6"
                  y1="6"
                  y2="18"
                />
                <line
                  x1="6"
                  x2="18"
                  y1="6"
                  y2="18"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </span>
  </div>
  <div>
    <button
      class="ventura button typePrimary"
      data-testid="button-toaster"
      name="toaster"
      type="button"
    >
      <span
        class="label"
      >
        Trigger toaster
      </span>
    </button>
  </div>
</body>
`;

exports[`Toaster snapshot of success toast 1`] = `
<body>
  <div
    id="ventura-toasters"
  >
    <span
      class="toastManagerContainer"
    >
      <div
        class="toastContainer"
        data-state="entering"
        style="height: 0px; margin-bottom: 0px;"
      >
        <div
          class="toastPadding"
        >
          <div
            class="ventura alert alertSuccess"
          >
            <div
              class="contentContainer closable"
            >
              <svg
                class="icon iconSuccess"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
                />
                <polyline
                  points="22 4 12 14.01 9 11.01"
                />
              </svg>
              <span
                class="message"
              >
                Message
              </span>
            </div>
            <div
              class="closeContainer"
            >
              <svg
                class="closeAlert"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="18"
                  x2="6"
                  y1="6"
                  y2="18"
                />
                <line
                  x1="6"
                  x2="18"
                  y1="6"
                  y2="18"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        class="toastContainer"
        data-state="entering"
        style="height: 0px; margin-bottom: 0px;"
      >
        <div
          class="toastPadding"
        >
          <div
            class="ventura alert alertError"
          >
            <div
              class="contentContainer closable"
            >
              <svg
                class="icon iconError"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="10"
                />
                <line
                  x1="12"
                  x2="12"
                  y1="8"
                  y2="12"
                />
                <line
                  x1="12"
                  x2="12.01"
                  y1="16"
                  y2="16"
                />
              </svg>
              <span
                class="message"
              >
                Message
              </span>
            </div>
            <div
              class="closeContainer"
            >
              <svg
                class="closeAlert"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="18"
                  x2="6"
                  y1="6"
                  y2="18"
                />
                <line
                  x1="6"
                  x2="18"
                  y1="6"
                  y2="18"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </span>
  </div>
  <div>
    <button
      class="ventura button typePrimary"
      data-testid="button-toaster"
      name="toaster"
      type="button"
    >
      <span
        class="label"
      >
        Trigger toaster
      </span>
    </button>
  </div>
</body>
`;
